main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
 object Inliner extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
-    * The maximum score of a function for it to be eligible for inlining.
+    * Performs inlining on the given AST `root`.
     */
-  val MaxScore = 25
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+    implicit val genSym: GenSym = flix.genSym
+
+    val definitions = root.definitions
+
+    val scores = definitions.foldLeft(Map.empty[DefnSym, Int]) {
+      case (macc, (sym, constant)) =>
+        macc + (sym -> Score.score(constant))
+    }
+
+    val inlinedDefinitions =
+    definitions.foldLeft(Map.empty[DefnSym, Constant]) {
+      case (macc, (sym,constant)) =>
+        val newExp = inlineExpr(definitions, scores, constant.exp)
+      macc + (sym -> constant.copy(exp = newExp))
+    }
+
+    root.copy(definitions = inlinedDefinitions).toSuccess
+
+  }
 
   /**
-    * Performs inlining on the given AST `root`.
+    * Performs the variable substitutions from `varSyms` into `exp`
     */
-  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+  def substituteWith(varSyms: Map[VarSym, VarSym], exp: Expression): Expression =
+  {
+    def f(e: Expression): Option[Expression] =
+    e match {
+      /* Substitute this expression */
+      case v: Var =>
+        varSyms.get(v.sym) match {
+          case None => None
+          case Some(sym) => Some(v.copy(sym = sym))
+        }
+      case MkClosureRef(ref, freeVars, tpe, loc) =>
+        Some(MkClosureRef(ref, freeVars.map(fv => varSyms.get(fv.sym) match {
+          case None => fv
+          case Some(sym) => FreeVar(sym, fv.tpe)
+        }), tpe, loc))
+      /* Substitute inside expression */
+      case _: LoadExpression => None
+      case _: StoreExpression => None
+      case Unit => None
```

Comment:
We usually "write out" one level of namespaces, so here we would write `Expression.Unit`. (This is because there is also `Type.Unit` and `Pattern.Unit` and we want to be able to see which one it is. (This goes not just for unit, but we prefix every expression node by well `Expression` and similar for Type and Pattern.)